rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,364 @@
+
+#### API
+
+The hook accepts a `MotionShorthand` param and a `MotionOptions`:
+
+```tsx
+// Types
+export type MotionType = 'unmounted' | 'entering' | 'entered' | 'idle' | 'exiting' | 'exited';
+
+export type MotionState<Element extends HTMLElement = HTMLElement> = {
+  /**
+   * Ref to the element.
+   */
+  ref: React.Ref<Element>;
+
+  /**
+   * Current state of the element.
+   *
+   * - `unmounted` - The element is not yet rendered or can be safely removed from the DOM.
+   * - `entering` - The element is performing enter animation.
+   * - `entered` - The element has finished enter animation.
+   * - `idle` - The element is currently not animating, but rendered on screen.
+   * - `exiting` - The element is performing exit animation.
+   * - `exited` - The element has finished exit animation.
+   */
+  type: MotionType;
+
+  /**
+   * Indicates whether the component is currently rendered and visible.
+   * Useful to apply CSS transitions only when the element is active.
+   */
+  isActive(): boolean;
+
+  /**
+   * Indicates whether the component can be rendered.
+   * This can be used to avoid rendering the component when it is not visible anymore.
+   */
+  canRender(): boolean;
+};
+
+export type MotionShorthandValue = boolean;
+
+export type MotionShorthand<Element extends HTMLElement = HTMLElement> = MotionShorthandValue | MotionState<Element>;
+
+type MotionOptions = {
+  /**
+   * Whether to animate the element on first mount. Useful when the animation/transition
+   * should be played if the element is already rendered on screen.
+   *
+   * @default false
+   */
+  animateOnFirstMount: false;
+};
+
+// Usage
+const [open, setOpen] = React.useState(false);
+const options = {
+  animateOnFirstMount: false,
+};
+const { ref, type, isActive, canRender } = useMotion(open, options);
+```
+
+The hook always returns a `MotionState`. The received `MotionShorthand` parameter can be either a `boolean` or a `MotionState`. This flexibility is extremely useful for cases when an override on another component is needed, and with that a double calculation is avoided. See the **Usage** section on this document.
+
+#### Usage
+
+##### State:
+
+```ts
+import * as React from 'react';
+import { getNativeElementProps, slot } from '@fluentui/react-utilities';
+import { useMotion } from '@fluentui/react-motion-preview';
+
+import type { SampleProps, SampleState } from './Sample.types';
+
+export const useSample_unstable = ({ open = false }: SampleProps, ref: React.Ref<HTMLElement>): SampleState => {
+  const motion = useMotion(open);
+
+  return {
+    components: {
+      root: 'div',
+    },
+
+    root: slot.always(
+      getNativeElementProps('div', {
+        ref: useMergedRefs(ref, motion.ref),
+        ...props,
+      }),
+      { elementType: 'div' },
+    ),
+
+    motion,
+  };
+};
+```
+
+##### Renderization:
+
+```tsx
+import * as React from 'react';
+import { getSlots } from '@fluentui/react-utilities';
```

Comment:
```suggestion
import { assertSlots } from '@fluentui/react-utilities';
```

Reply:
Just to keep it up to date with the current documentation 😁 https://react.fluentui.dev/?path=/docs/concepts-developer-customizing-components-with-slots--page#rendering-components-with-slots